pkg/util/resource_parent_map.go

```diff
@@ -0,0 +1,220 @@
+
+import (
+	"fmt"
+	"sync"
+
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+)
+
+var log = logf.Log.WithName("controller")
+
+// Kind strings to be used in KubeResource object creation
+// Use these kinds when adding to the ResourceParentsMap
+const (
+	KindSecret    = "Secret"
+	KindNamespace = "Namespace"
+
+	KindMigPlan            = "MigPlan"
+	KindMigCluster         = "MigCluster"
+	KindMigAssetCollection = "MigAssetCollection"
+	KindMigStorage         = "MigStorage"
+
+	KindMigStage     = "MigStage"
+	KindMigMigration = "MigMigration"
+
+	KindClusterRegCluster = "Cluster"
+
+	KindVeleroBackup                 = "VeleroBackup"
+	KindVeleroRestore                = "VeleroRestore"
+	KindVeleroBackupStorageLocation  = "VeleroBackupStorageLocation"
+	KindVeleroVolumeSnapshotLocation = "VeleroVolumeSnapshotLocation"
+)
+
+var rpmInstance *ResourceParentsMap
+var createChildParentsMapOnce sync.Once
+
+// KubeResource is a simplified way to related child to parent resources
+type KubeResource struct {
+	Kind   string
+	NsName types.NamespacedName
+}
+
+// ResourceParentsMap maps from "parents resources" to "child resources"
+// Maintaining this allows event triggering with a 1:N reference relationship
+type ResourceParentsMap struct {
+	sync.RWMutex
+	childToParentsMap map[KubeResource][]KubeResource // 1:N child:parent mapping
+}
+
+// GetResourceParentsMap must be called to get a reference to the singleton ResourceParentsMap
+func GetResourceParentsMap() *ResourceParentsMap {
+	createChildParentsMapOnce.Do(func() {
+		rpmInstance = &ResourceParentsMap{}
+		rpmInstance.childToParentsMap = make(map[KubeResource][]KubeResource)
+	})
+	return rpmInstance
+}
+
+// Definitely useful
+//  - [DONE] Get the list of parent resources _of a particular kind_ for a particular child resource
+//  - [DONE] Add a reference from a child resource to a parent resource
+//  - [DONE] Delete a reference from a child resource to a parent resource
+//  - [DONE] Get list of child resources for a particular parent
+
+// Maybe useful
+//  - Get the list of all child resources of a particular kind
+
+// GetParentsOfKind ...
+func (r *ResourceParentsMap) GetParentsOfKind(child KubeResource, parentKind string) []KubeResource {
+	r.RLock()
+	defer r.RUnlock()
+
+	parents, ok := rpmInstance.childToParentsMap[child]
+	if !ok {
+		return nil
+	}
+
+	// Find parents of particular kind
+	matchingParents := []KubeResource{}
+	for i := range parents {
+		if parents[i].Kind == parentKind {
+			matchingParents = append(matchingParents, parents[i])
+		}
+	}
+
+	// Will return empty slice if nothing found
+	return matchingParents
+}
+
+// GetChildrenOfKind ...
+// WARNING: this function has not been tested. It is likely to not work as desired.
+func (r *ResourceParentsMap) GetChildrenOfKind(parent KubeResource, childKind string) []KubeResource {
```

Comment:
I'm assuming this is intended to support cleaning the map when something _referenced_ is deleted.  I think it would be better to only update the map when a reference changes and this method should be removed.

Reply:
ACK, per our realization that we will likely be able to manage these mappings from predicate, I agree. I was trying to think through all of the ways it would be necessary to pull data out of the map and thought this may come in handy, but turned out to not use it anywhere so far.